[libc] Add base for target config within cmake

michaelrj-google opened the pull request:
Currently the only way to add or remove entrypoints is to modify the
entrypoints.txt file for the current target. This isn't ideal since
a user would have to carry a diff for this file when updating their
checkout. This patch adds a basic mechanism to allow the user to remove
entrypoints without modifying the repository.

michaelrj-google committed:
[libc] Add base for target config within cmake

Currently the only way to add or remove entrypoints is to modify the
entrypoints.txt file for the current target. This isn't ideal since
a user would have to carry a diff for this file when updating their
checkout. This patch adds a basic mechanism to allow the user to remove
entrypoints without modifying the repository.

michaelrj-google committed:
Move to completely overwrite the entrypoints path

Instead of having a user-provided file that modifies the list of
entrypoints, now the user-provided file is used instead of the list of
entrypoints.

lntue commented on libc/CMakeLists.txt:
```diff
@@ -249,40 +249,43 @@ include(CMakeParseArguments)
+#   list(REMOVE_ITEM TARGET_LIBC_ENTRYPOINTS ${removed_entrypoint})
+#   list(REMOVE_ITEM TARGET_LIBM_ENTRYPOINTS ${removed_entrypoint})
+# endforeach()
+
```
WDYT of the following reorg:
```
if(NOT LIBC_CONFIG_PATH)
  if(EXISTS "${LIBC_SOURCE_DIR}/config/${LIBC_TARGET_OS}/${LIBC_TARGET_ARCHITECTURE}")
    set(LIBC_CONFIG_PATH "${LIBC_SOURCE_DIR}/config/${LIBC_TARGET_OS}/${LIBC_TARGET_ARCHITECTURE}")
  elseif(EXISTS "${LIBC_SOURCE_DIR}/config/${LIBC_TARGET_OS}")
    set(LIBC_CONFIG_PATH "${LIBC_SOURCE_DIR}/config/${LIBC_TARGET_OS}")
endif()

if(NOT LIBC_CONFIG_PATH)
  message(FATAL_ERROR "Configs for the platform '${LIBC_TARGET_OS}/${LIBC_TARGET_ARCHITECTURE}' do not exist and LIBC_CONFIG_PATH is not set.")
elseif(LIBC_CMAKE_VERBOSE_LOGGING)
  message(STATUS "Path for config files is: ${LIBC_CONFIG_PATH}")
endif()

# Check entrypoints.txt
if(EXISTS "${LIBC_CONFIG_PATH}/entrypoints.txt")
    set(entrypoint_file "${LIBC_CONFIG_PATH}/entrypoints.txt")
else()
  message(FATAL_ERROR "${LIBC_CONFIG_PATH}/entrypoints.txt file not found.")
endif()

# Check header.txt
if(EXISTS "${LIBC_CONFIG_PATH}/headers.txt")
    include("${LIBC_CONFIG_PATH}/headers.txt")
endif()

# Check exclude.txt that appends to LIBC_EXCLUDE_ENTRYPOINTS list
if(EXISTS "${LIBC_CONFIG_PATH}/exclude.txt")
    include("${LIBC_CONFIG_PATH}/exclude.txt")
endif()

// Remove entry points:
foreach(removed_entrypoint IN LISTS LIBC_EXCLUDE_ENTRYPOINTS)
   ...
```

michaelrj-google replied:
I did something similar to this, though I had to reorder it a bunch to handle the config files.

michaelrj-google committed:
Redesign to support providing a config folder

Instead of the user only providing one file, they now provide a file
with entrypoints, headers, and the new exclude file. It also now
supports them providing their own config.json and properly uses it
instead of the provided ones for the given target.

lntue commented on libc/config/linux/x86_64/exclude.txt:
```diff
@@ -0,0 +1,16 @@
+# This optional file is used to exclude entrypoints/headers for specific targets.
+
+try_compile(
+  has_sys_random
+  ${CMAKE_CURRENT_BINARY_DIR}
+  SOURCES ${LIBC_SOURCE_DIR}/src/sys/random/linux/getrandom.cpp
+)
```
We should only do this in overlay mode right?  Since we do know whether the header exists or not in full build mode.

lntue reviewed (commented):
So currently we (potentially) include multiple `config.json` files.  I think each build should only has 1 single `config.json`, similar to `entrypoints.txt`, and should be put in the same path.  If multiple architectures on the same OS want to share some configs, then they should explicitly include the shared `config.json` inside their own `config.json`.

Another the thing we should do is probably refactor all of these logic into `LibcConfig` file.

All of these improvements can be done in the followup changes.  This PR is good to me as it is to enable the feature.

michaelrj-google committed:
Properly set up exclude.txt

I had to add a new folder for files to use try_compile on.
